### ISSUE TITLE
__syscall4 being called infinitely causing Uncaught RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by iamtekeste** [emscripten-core/emscripten#10884]
Can anyone point me in the direction to solve this issue? I am guessing this has something to do with file access/file system.

```js
Uncaught RangeError: Maximum call stack size exceeded
    at ___syscall4 (pdftocairo.js:4989)
    at __sanitizer::internal_write(int, void const*, unsigned long) (:5000/wasm-function[13869]:0x8d67b8)
    at __sanitizer::IsAccessibleMemoryRange(unsigned long, unsigned long) (:5000/wasm-function[14028]:0x8db199)
    at __ubsan::checkDynamicType(void*, void*, unsigned long) (:5000/wasm-function[13792]:0x8d1d65)
    at HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) (:5000/wasm-function[13761]:0x8cefda)
    at __ubsan_handle_dynamic_type_cache_miss (:5000/wasm-function[13760]:0x8cef9a)
    at std::type_info::operator==(std::type_info const&) const (:5000/wasm-function[8873]:0x6644bd)
    at is_equal(std::type_info const*, std::type_info const*, bool) (:5000/wasm-function[13691]:0x8ca3d3)
    at __dynamic_cast (:5000/wasm-function[13694]:0x8ca52b)
    at __ubsan::checkDynamicType(void*, void*, unsigned long) (:5000/wasm-function[13792]:0x8d1d9a)
```
I have put some console.log inside it and it printed `2242` times.
Here is the body of that method
```js
  function ___syscall4(fd, buf, count) {try {
   // write
      var stream = SYSCALLS.getStreamFromFD(fd);
      return FS.write(stream, HEAP8,buf, count);
    } catch (e) {
    if (typeof FS === 'undefined' || !(e instanceof FS.ErrnoError)) abort(e);
    return -e.errno;
  }
  }
```

**Comment by sbc100** [comment]
It looks like you are trying to do a ubsen build here?  (`-fsanitize=undefined`?)  Does you program work without the sanitizer enabled?

Presumably the stack trace you are including is not complete?  Can you include more of that stack strace so we can see where if the recursion is happening?

 I'm surprised to see that ubsan is depending on `pipe` and `write` (see `sanitizer_posix_libcdep.cc:HandleDynamicTypeCacheMiss`).    

I imagine this might be related to the filesystem you are you using.  What link flags are you using?  (In particular what filesystem are you enabling?)

**Comment by iamtekeste** [comment]
> It looks like you are trying to do a ubsen build here? (-fsanitize=undefined?) Does you program work without the sanitizer enabled?

It does not work without sanitizer enabled, I added it to debug the error I got. Which was the following and has more of the stack trace

```js
pdftocairo.js:41 Uncaught RuntimeError: function signature mismatch
    at _cairo_spline_add_point (wasm-function[1938]:0x3ecdd)
    at _cairo_spline_decompose_into (wasm-function[1936]:0x3eaf3)
    at _cairo_spline_decompose_into (wasm-function[1936]:0x3eb15)
    at _cairo_spline_decompose_into (wasm-function[1936]:0x3eb08)
    at _cairo_spline_decompose_into (wasm-function[1936]:0x3eb08)
    at _cairo_spline_decompose (wasm-function[1935]:0x3ea84)
    at _cpf_curve_to (wasm-function[1326]:0x27e14)
    at _cairo_path_fixed_interpret (wasm-function[1310]:0x27608)
    at _cairo_path_fixed_interpret_flat (wasm-function[1323]:0x27d66)
    at _cairo_clip_reduce_to_boxes (wasm-function[735]:0x19480)

  | _cairo_spline_add_point | @ | CairoRescaleBox.h:49
  | _cairo_spline_decompose_into | @ | CairoRescaleBox.h:49
  | _cairo_spline_decompose_into | @ | CairoRescaleBox.h:49
  | _cairo_spline_decompose_into | @ | CairoRescaleBox.h:49
  | _cairo_spline_decompose_into | @ | CairoRescaleBox.h:49
  | _cairo_spline_decompose | @ | CairoRescaleBox.h:49
  | _cpf_curve_to | @ | CairoRescaleBox.h:49
  | _cairo_path_fixed_interpret | @ | CairoRescaleBox.h:49
  | _cairo_path_fixed_interpret_flat | @ | CairoRescaleBox.h:49
  | _cairo_clip_reduce_to_boxes | @ | CairoRescaleBox.h:49
  | _cairo_clip_reduce_to_rectangle | @ | CairoRescaleBox.h:49
  | _cairo_clip_reduce_for_composite | @ | CairoRescaleBox.h:49
  | _cairo_composite_rectangles_init_for_paint | @ | CairoRescaleBox.h:49
  | _cairo_recording_surface_paint | @ | CairoRescaleBox.h:49
  | _cairo_surface_paint | @ | CairoRescaleBox.h:49
  | _cairo_paginated_surface_paint | @ | CairoRescaleBox.h:49
  | _cairo_surface_paint | @ | CairoRescaleBox.h:49
  | _cairo_gstate_fill | @ | CairoRescaleBox.h:49
  | _cairo_default_context_fill | @ | CairoRescaleBox.h:49
  | cairo_fill | @ | CairoRescaleBox.h:49
  | CairoOutputDev::tilingPatternFill(GfxState*, Gfx*, Catalog*, Object*, double const*, int, int, Dict*, double const*, double const*, int, int, int, int, double, double) | @ | CairoOutputDev.cc:982
  | Gfx::doTilingPatternFill(GfxTilingPattern*, bool, bool, bool) | @ | CairoRescaleBox.h:49
  | Gfx::doPatternFill(bool) | @ | CairoRescaleBox.h:49
  | Gfx::opFill(Object*, int) | @ | CairoRescaleBox.h:49
  | Gfx::execOp(Object*, Object*, int) | @ | CairoRescaleBox.h:49
  | Gfx::go(bool) | @ | CairoRescaleBox.h:49
  | Gfx::display(Object*, bool) | @ | CairoRescaleBox.h:49
  | Page::displaySlice(OutputDev*, double, double, int, bool, bool, int, int, int, int, bool, bool (*)(void*), void*, bool (*)(Annot*, void*), void*, bool) | @ | CairoRescaleBox.h:49
  | PDFDoc::displayPageSlice(OutputDev*, int, double, double, int, bool, bool, bool, int, int, int, int, bool (*)(void*), void*, bool (*)(Annot*, void*), void*, bool) | @ | CairoRescaleBox.h:49
  | renderPage(PDFDoc*, CairoOutputDev*, int, double, double, double, double) | @ | pdftocairo.cc:736
  | main | @ | pdftocairo.cc:1268
  | Module._main | @ | pdftocairo.js:5827
  | callMain | @ | pdftocairo.js:6770
  | reader.onload | @ | (index):58
  | load (async) |   |  
  | (anonymous)
```
This error occurs with some PDF files only, which I have attached here.
[bad pdf.pdf](https://github.com/emscripten-core/emscripten/files/4465798/bad.pdf.pdf)

> I imagine this might be related to the filesystem you are you using. What link flags are you using? (In particular what filesystem are you enabling?)

I didn't specifically enable any file system but the code I am trying to port does make calls to `fopen` and I use `FS.writeFile` in the JavaScript side of things to make files available to it.

Here is the source code for the project I am trying to port
https://gitlab.freedesktop.org/poppler/poppler
Specifically, the `pdftocairo` utility located here, https://gitlab.freedesktop.org/poppler/poppler/-/blob/master/utils/pdftocairo.cc

If it is helpful I have also reduced version of pdftocairo.cc that only does PDF to SVG conversion
https://gist.github.com/iamtekeste/847f91ee46317ab8fb7ca523f8fdf049

As you might have guessed `pdftocairo` depends on` Cairo` library which I was able to port although there was a bunch of warnings regarding implicit enumeration conversion.

Let me know if you want to take a look at the generated JavaScript and WASM files too.

**Comment by kripken** [comment]
Does `Maximum call stack size exceeded` (or similar) occur on different VMs? If so it might not be a VM limit but an infinite recursion. But if that's the full stack trace then perhaps that's not the issue as it doesn't seem recursive there.

Can confirm by printing the stack trace in those console.logs in that function.

(If this isn't an infinite recursion, then it's a VM limit I assume.)

**Comment by iamtekeste** [comment]
I am a beginner to the WebAssembly world, does VM mean another browser in this case? If yes, then yeah I see the issue on Firefox too.

I don't really think it is a recursion issue since the native compiled/built native code works great.

After building with `-fsanitize=undefined ` and putting `console.trace` on that syscall method, I got the following stack trace. @kripken 

```js

  | ___syscall42 | @ | pdftocairo.js:5174
  | pipe | @ | memory:2245
  | __sanitizer::IsAccessibleMemoryRange(unsigned long, unsigned long) | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | PDFDocFactory::createPDFDoc(GooString const&, GooString*, GooString*, void*) | @ | PDFDocFactory.cc:62
  | main | @ | pdftocairo.cc:1143
  | Module._main | @ | pdftocairo.js:6605
  | callMain | @ | pdftocairo.js:7569
  | reader.onload | @ | (index):58
  | load (async) |   |  
  | (anonymous)
```

**Comment by kripken** [comment]
Ok, great, then that stack trace confirms it's an infinite recursion. So somewhere here we need to break the loop, inside the ubsan code:
```js
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | std::type_info::operator==(std::type_info const&) const | @ | typeinfo:306
  | is_equal(std::type_info const*, std::type_info const*, bool) | @ | memory:2245
```
I don't know that much about ubsan but this may be one of the things where the sanitizer needs to either call a special function (I think we have a special malloc for it, for example), or where it needs to detect the recursion and stop it.

This all begins with
```js
  | __dynamic_cast | @ | memory:2245
  | __ubsan::checkDynamicType(void*, void*, unsigned long) | @ | memory:2245
  | HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) | @ | memory:2245
  | __ubsan_handle_dynamic_type_cache_miss | @ | memory:2245
  | PDFDocFactory::createPDFDoc(GooString const&, GooString*, GooString*, void*) | @ | PDFDocFactory.cc:62
```
(where the PDF code is still normal user code, and everything above it is ubsan and what ubsan calls). From that, my guess is that ubsan needs to call a special *unchecked* `dynamic_cast` that won't infinitely recurse by calling back into ubsan for more checks. But that's just a guess.

cc @tlively

**Comment by iamtekeste** [comment]
Is there anything I can do from my side to get this sorted out? Or is it a bug on Emscripten side?

cc: @sbc100 @kripken @tlively

**Comment by kripken** [comment]
I think we are all in the same situation here, basically - none of us really know the UBSan code. I'm not sure if this is a bug in UBSan or in emscripten's integration with it. Either way, looking at the UBSan code in the emscripten repo (somewhere under `lib/`) is the right place to start.

Oh, to check if this is a general bug in UBSan you can perhaps try to see if it reproduces in a non-emscripten build.

**Comment by tlively** [comment]
I'm not too familiar with the UBSan code myself, but I can take a look on Monday.

**Comment by iamtekeste** [comment]
Thank you all for your help!

**Comment by jamesruan** [comment]
I also run a cross the same problem when I try to enable UBSan:
`canvas.js:4506 Uncaught (in promise) RangeError: Maximum call stack size exceeded
    at ___syscall4 (canvas.js:4506)
    at __sanitizer::internal_write(int, void const*, unsigned long) (:8080/wasm-function[17284]:0x5b3a93)
    at __sanitizer::IsAccessibleMemoryRange(unsigned long, unsigned long) (:8080/wasm-function[17325]:0x5b4e6a)
    at __ubsan::checkDynamicType(void*, void*, unsigned long) (:8080/wasm-function[17168]:0x5accde)
    at HandleDynamicTypeCacheMiss(__ubsan::DynamicTypeCacheMissData*, unsigned long, unsigned long, __ubsan::ReportOptions) (:8080/wasm-function[17182]:0x5ad6f1)
    at __ubsan_handle_dynamic_type_cache_miss (:8080/wasm-function[17181]:0x5ad6c9)
    at std::type_info::operator==(std::type_info const&) const (:8080/wasm-function[2563]:0x94472)
    at is_equal(std::type_info const*, std::type_info const*, bool) (:8080/wasm-function[17070]:0x5a534d)
    at __dynamic_cast (:8080/wasm-function[17072]:0x5a547b)
    at __ubsan::checkDynamicType(void*, void*, unsigned long) (:8080/wasm-function[17168]:0x5acd12)`

**Comment by tlively** [comment]
@iamtekeste Sorry it's taken so long, but I'm trying to look into this now. I think I know how to fix it just by looking at the UBSan code, but I wasn't able to build freetype to reproduce the issue so I can't check. Can you share the exact steps you take to build the project and reproduce the issue? @jamesruan if you can share reproduction steps I can take a look at your version of the problem as well.

**Comment by iamtekeste** [comment]
No worries @tlively Here is the modified Cmake file. Most of the change is me adding 
`set(FREETYPE_ME "-s USE_FREETYPE=1")`which means I wouldn't need to build freetype.
https://gist.github.com/iamtekeste/d73b70d5085099e7a0d39b2ac78c6917

If you can share what the fix might be I can also give it a shot on my machine.

**Comment by jamesruan** [comment]
@tlively Sorry. I was building a private project based on Skia, not easy to share. I was going to let UBSan help me find a potention memory problem, enable it with `-fsanitiser=undefined` produce above error.

**Comment by tlively** [comment]
@iamtekeste What commit from https://gitlab.freedesktop.org/poppler/poppler are you using? Also, do you have instructions for getting your cairo port? cmake is failing with your modified cmake file without it.

**Comment by tlively** [comment]
This should have been fixed by #13367.
